python/ffsim/hamiltonians/diagonal_coulomb_hamiltonian.py

```diff
@@ -0,0 +1,171 @@
+        """Return a SciPy LinearOperator representing the object."""
+        dim_ = dim(norb, nelec)
+        eigs, vecs = scipy.linalg.eigh(self.one_body_tensor)
+        num_linop = num_op_sum_linop(eigs, norb, nelec, orbital_rotation=vecs)
+        dc_linop = diag_coulomb_linop(
+            (
+                self.diag_coulomb_mats[0],
+                self.diag_coulomb_mats[1],
+                self.diag_coulomb_mats[0],
+            ),
+            norb,
+            nelec,
+        )
+
+        def matvec(vec: np.ndarray):
+            result = self.constant * vec
+            result += num_linop @ vec
+            result += dc_linop @ vec
+            return result
+
+        return LinearOperator(
+            shape=(dim_, dim_), matvec=matvec, rmatvec=matvec, dtype=complex
+        )
+
+    def _fermion_operator_(self) -> FermionOperator:
+        """Return a FermionOperator representing the object."""
+        op = FermionOperator({(): self.constant})
+        for p, q in itertools.product(range(self.norb), repeat=2):
+            coeff = self.one_body_tensor[p, q]
+            op += FermionOperator(
+                {
+                    (cre_a(p), des_a(q)): coeff,
+                    (cre_b(p), des_b(q)): coeff,
+                }
+            )
+        for p, q in itertools.product(range(self.norb), repeat=2):
+            (coeff_same_spin, coeff_diff_spin) = 0.5 * self.diag_coulomb_mats[:, p, q]
+            op += FermionOperator(
+                {
+                    (cre_a(p), des_a(p), cre_a(q), des_a(q)): coeff_same_spin,
+                    (cre_a(p), des_a(p), cre_b(q), des_b(q)): coeff_diff_spin,
+                    (cre_b(p), des_b(p), cre_a(q), des_a(q)): coeff_diff_spin,
+                    (cre_b(p), des_b(p), cre_b(q), des_b(q)): coeff_same_spin,
+                }
+            )
+        return op
+
+    @staticmethod
+    def from_fermion_operator(op: FermionOperator) -> DiagonalCoulombHamiltonian:
+        """Convert a FermionOperator to a DiagonalCoulombHamiltonian."""
+
+        # extract norb
+        norb = 1 + max(orb for term in op for _, _, orb in term)
+
+        # initialize variables
+        constant: float = 0
+        one_body_tensor = np.zeros((norb, norb), dtype=complex)
+        diag_coulomb_mats = np.zeros((2, norb, norb), dtype=float)
+
+        # populate the tensors
+        for term, coeff in op.items():
+            if not term:  # constant term
+                constant = coeff.real
+            elif len(term) == 2:  # one-body term candidate
+                (_, _, p), (_, _, q) = term
+                terms = [(cre_a(p), des_a(q)), (cre_b(p), des_b(q))]
+                if term in terms:
+                    one_body_tensor[p, q] += 0.5 * coeff
+                else:
+                    raise ValueError(
+                        "FermionOperator cannot be converted to "
+                        "DiagonalCoulombHamiltonian"
```

Comment:
Make this error message more descriptive. Explain that it encountered a quadratic term that was not of the form `a^\dagger_p a_q`, and include the term in the error message (use an f-string).